File: Alkitab/src/yuku/alkitab/base/storage/YesReaderFactory.java
package yuku.alkitab.base.storage;

import android.util.Log;

import java.io.IOException;
import java.io.RandomAccessFile;
import java.util.Arrays;

import yuku.alkitab.yes1.Yes1Reader;
import yuku.alkitab.yes2.Yes2Reader;
import yuku.alkitab.yes2.io.RandomInputStream;

public class YesReaderFactory {
	public static final String TAG = YesReaderFactory.class.getSimpleName();
	
	/**
	 * @return A {@link Yes1Reader}, {@link Yes2Reader}, or null if there is any error.
	 */
	public static BibleReader createYesReader(String filename) {
		try {
			RandomAccessFile f = new RandomAccessFile(filename, "r");
			byte[] header = new byte[8];
			f.read(header);
			
			if (header[0] != (byte) 0x98 
			|| header[1] != (byte) 0x58 
			|| header[2] != (byte) 0x0d 
			|| header[3] != (byte) 0x0a 
			|| header[4] != (byte) 0x00 
			|| header[5] != (byte) 0x5d 
			|| header[6] != (byte) 0xe0) {
				Log.e(TAG, "Yes file has not a correct header. Header is: " + Arrays.toString(header));
				return null;
			}
			
			if (header[7] == 0x01) { // VERSION 1 YES
				return new Yes1Reader(f);
			} else if (header[7] == 0x02) { // VERSION 2 YES
				return new Yes2Reader(new RandomInputStream(f));
			} else {
				Log.e(TAG, "Yes file version unsupported: " + header[7]);
				return null;
			}
		} catch (IOException e) {
			Log.e(TAG, "@@createYesReader io exception", e);
			return null;
		}
	}
}

File: Alkitab/src/main/java/yuku/alkitab/base/storage/YesReaderFactory.kt
package yuku.alkitab.base.storage

import java.io.IOException
import java.io.RandomAccessFile
import yuku.alkitab.base.util.AppLog
import yuku.alkitab.io.BibleReader
import yuku.alkitab.yes1.Yes1Reader
import yuku.alkitab.yes2.Yes2Reader
import yuku.alkitab.yes2.io.RandomAccessFileRandomInputStream

object YesReaderFactory {
    private val TAG: String = YesReaderFactory::class.java.simpleName

    private val MAGIC = byteArrayOf(0x98.toByte(), 0x58, 0x0d, 0x0a, 0x00, 0x5d, 0xe0.toByte())

    /**
     * @return A [Yes1Reader], [Yes2Reader], or null if there is any error.
     */
    @JvmStatic
    fun createYesReader(filename: String): BibleReader? {
        try {
            val header = ByteArray(8)

            val f = RandomAccessFile(filename, "r")
            f.read(header)
            f.close()

            if (!header.copyOfRange(0, MAGIC.size).contentEquals(MAGIC)) {
                AppLog.e(TAG, "Yes file '" + filename + "' has not a correct header. Header is: " + header.contentToString())
                return null
            }

            return when (header[7].toInt()) {
                0x01 -> Yes1Reader(filename) // VERSION 1 YES
                0x02 -> Yes2Reader(RandomAccessFileRandomInputStream(filename)) // VERSION 2 YES
                else -> {
                    AppLog.e(TAG, "Yes file version unsupported: " + header[7])
                    null
                }
            }
        } catch (e: IOException) {
            AppLog.e(TAG, "@@createYesReader io exception", e)
            return null
        }
    }
}
